File: signals/signals.js
const { print, net } = just
const { signal } = just.library('signal', './modules/signal/signal.so')
const { EPOLLIN } = just.loop
const { signalfd, sigaddset, sigprocmask, sigemptyset } = signal
const { loop } = just.factory

const stringify = (o, sp = '  ') => JSON.stringify(o, (k, v) => (typeof v === 'bigint') ? v.toString() : v, sp)

function onSignal (siginfo) {
  const view = new DataView(siginfo)
  const signo = view.getUint32(0, true)
  const errno = view.getInt32(4, true)
  const code = view.getInt32(8, true)
  const pid = view.getUint32(12, true)
  const uid = view.getUint32(16, true)
  const fd = view.getInt32(20, true)
  const tid = view.getUint32(24, true)
  const band = view.getUint32(28, true)
  const overrun = view.getUint32(32, true)
  const trapno = view.getUint32(36, true)
  const status = view.getInt32(40, true)
  const int = view.getInt32(44, true)
  const ptr = view.getBigUint64(48, true)
  const utime = view.getBigUint64(56, true)
  const stime = view.getBigUint64(64, true)
  const addr = view.getBigUint64(72, true)
  print(stringify({ signo, errno, code, pid, uid, fd, tid, band, overrun, trapno, status, int, ptr, utime, stime, addr }))
}

const sigmask = new ArrayBuffer(1024)
const siginfo = new ArrayBuffer(1024)

let r = 0

signal.reset()

r = sigprocmask(sigmask, signal.SIG_SETMASK, 1)
if (r < 0) throw new SystemError('sigprocmask')
r = sigaddset(sigmask, signal.SIGUSR1)
if (r < 0) throw new SystemError('sigaddset')
r = sigaddset(sigmask, signal.SIGHUP)
if (r < 0) throw new SystemError('sigaddset')
r = sigaddset(sigmask, signal.SIGINT)
if (r < 0) throw new SystemError('sigaddset')
r = sigprocmask(sigmask, signal.SIG_SETMASK, 0)
if (r < 0) throw new SystemError('sigprocmask')

r = sigemptyset(sigmask)
if (r < 0) throw new SystemError('sigemptyset')
r = sigaddset(sigmask, signal.SIGUSR1)
if (r < 0) throw new SystemError('sigaddset')
r = sigaddset(sigmask, signal.SIGHUP)
if (r < 0) throw new SystemError('sigaddset')
r = sigaddset(sigmask, signal.SIGINT)
if (r < 0) throw new SystemError('sigaddset')

const sigfd = signalfd(sigmask, 0)
if (sigfd <= 0) throw new SystemError('signalfd')

const pid = just.sys.pid()

r = loop.add(sigfd, (fd, event) => {
  just.print('event')
  if (event & EPOLLIN) {
    just.print('EPOLLIN')
    let bytes = net.read(fd, siginfo)
    just.print(bytes)
    if (bytes < 0) {
      //net.close(fd)
      return
    }
    onSignal(siginfo)
  }
}, EPOLLIN)
if (r < 0) throw new SystemError('loop.add')

just.setInterval(() => {
  just.print(`pid ${pid} mem ${just.memoryUsage().rss}`)
}, 1000)

File: signals/signals.c
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/epoll.h>
#include <signal.h>
#include <unistd.h>
#include <sys/signalfd.h>
#include <inttypes.h>
#include <stdarg.h>

static void die(char *s, ...) {
	va_list v;

	va_start(v, s);
	vfprintf(stderr, s, v);
	fprintf(stderr, "\n");
	va_end(v);
	fprintf(stderr, " -- errno = %i (%m)\n", errno);

	fflush(stderr);
	abort();
}

static void print_event (struct epoll_event ev) {
	char flags_str[200];

	flags_str[0] = 0;
	flags_str[1] = 0; /* dirty */

	if (ev.events & EPOLLIN )	strcat(flags_str, "|EPOLLIN");
	if (ev.events & EPOLLOUT)	strcat(flags_str, "|EPOLLOUT");
	if (ev.events & EPOLLERR)	strcat(flags_str, "|EPOLLERR");
	if (ev.events & EPOLLHUP)	strcat(flags_str, "|EPOLLHUP");
	if (ev.events & EPOLLRDHUP)	strcat(flags_str, "|EPOLLRDHUP");
	if (ev.events & EPOLLPRI)	strcat(flags_str, "|EPOLLPRI");
	if (ev.events & EPOLLET)	strcat(flags_str, "|EPOLLET");
	if (ev.events & EPOLLONESHOT)	strcat(flags_str, "|EPOLLONESHOT");

	fprintf(stderr, "Event for fd %i. Flags=(%s).\n",
			ev.data.fd, flags_str+1);
}

static void signal_handler (int signum) {
  fprintf(stderr, "signal %i\n", signum);
}

int setup_signalfd() {
	int sfd, ret;
	sigset_t sigset;

	ret = sigprocmask(SIG_SETMASK, NULL, &sigset);
	if (ret < 0)
		die("sigprocmask.1");

	sigaddset(&sigset, SIGINT);
	sigaddset(&sigset, SIGUSR1);
	sigprocmask(SIG_SETMASK, &sigset, NULL);
	if (ret < 0)
		die("sigprocmask.2");

	sigemptyset(&sigset);
	sigaddset(&sigset, SIGINT);
	sigaddset(&sigset, SIGUSR1);

	sfd = signalfd(-1, &sigset, 0);
	if (sfd < 0)
		die("signalfd");

	printf("sfd is %i\n", sfd);
	return sfd;
}

void read_sig(int sfd) {
	struct signalfd_siginfo info;
	int ret;

	ret = read(sfd, &info, sizeof info);
	if (ret != sizeof info)
		die("!?!?!?!");

	printf("Signal Received\n");
	printf("signo = %" PRIu32 "\n", info.ssi_signo);
	printf("pid   = %" PRIu32 "\n", info.ssi_pid);
	printf("uid   = %" PRIu32 "\n", info.ssi_uid);
}

int main() {
	int epfd;
	struct epoll_event ev;
	struct epoll_event ret;
	char buf[200];
	int n, t;
	int sfd;

	epfd = epoll_create(1);
	sfd = setup_signalfd();
	ev.data.fd = sfd;
	ev.events = EPOLLIN;
	if (epoll_ctl(epfd, EPOLL_CTL_ADD, sfd, &ev) != 0)
		perror("epoll_ctl");

	while ((n = epoll_wait(epfd, &ret, 1, -1)) > 0) {
		printf("tick!\n");
		print_event(ret);

		if (ret.data.fd == 0) {
			t = read(0, buf, 100);

			if (t == 0) {
				epoll_ctl(epfd, EPOLL_CTL_DEL, 0, NULL);
				close(0);
				printf("stdin done\n");
			}
		} else if (ret.data.fd == sfd) {
			read_sig(sfd);
		}
	}

	return 0;
}